core/src/main/java/org/apache/accumulo/core/client/admin/InstanceOperations.java

```diff
@@ -181,41 +184,90 @@ Map<String,String> modifyProperties(Consumer<Map<String,String>> mapMutator)
    *
    * @return a list of locations in <code>hostname:port</code> form.
    * @since 2.1.0
+   * @deprecated see {@link #getServers(ServerTypeName)}
    */
+  @Deprecated(since = "4.0.0")
   List<String> getManagerLocations();
 
   /**
    * Returns the locations of the active compactors
    *
    * @return A set of currently active compactors.
    * @since 2.1.4
+   * @deprecated see {@link #getServers(ServerTypeName)}
    */
+  @Deprecated(since = "4.0.0")
   Set<String> getCompactors();
 
   /**
    * Returns the locations of the active scan servers
    *
    * @return A set of currently active scan servers.
+   * @deprecated see {@link #getServers(ServerTypeName)}
    */
+  @Deprecated(since = "4.0.0")
   Set<String> getScanServers();
 
   /**
    * List the currently active tablet servers participating in the accumulo instance
    *
    * @return A list of currently active tablet servers.
+   * @deprecated see {@link #getServers(ServerTypeName)}
    */
+  @Deprecated(since = "4.0.0")
   List<String> getTabletServers();
 
+  /**
+   * Resolve the server of the given type and address to a ServerId
+   *
+   * @param type type of server
+   * @param host host name
+   * @param port host port
+   * @return ServerId if found, else null
+   * @since 4.0.0
+   */
+  ServerId getServer(ServerTypeName type, String host, int port);
+
+  /**
+   * Returns all servers of the given types
+   *
+   * @return set of servers of the supplied types matching the supplied test
+   * @since 4.0.0
+   */
+  Set<ServerId> getServers(ServerTypeName type);
+
+  /**
+   * Returns the servers of a given type that match the given criteria
+   *
+   * @return set of servers of the supplied types matching the supplied test
+   * @since 4.0.0
+   */
+  Set<ServerId> getServers(ServerTypeName type, Predicate<ServerId> test);
```

Comment:
I think to do this efficiently, the resource group would need to be in the zookeeper path for all of the server processes (well, the ones that use them). Right now, the resource group is only in the path of the Compactor process. I saw your comment that you didn't look too deeply at the ServerIds class, but that would show you the different paths in use.

We could modify the TabletServer and ScanServer zookeeper paths in a separate PR to align with the Compactor ZK path now that the group is mandatory. That would simplify the ServerIds class and allow for your suggestion to be implemented efficiently.

Reply:
> Maybe we could first do a PR to add RG filtering to listscans and listcompactions in the shell and then later PRs to optimize the filtering.

Issue #4782 has been submitted to add filtering to list compactions. I"m not sure which version that solution will target. IMO, this PR is an optimization for that filtering, specifically for the 4.0 version. I think there could be further optimization done, as you suggested, in another PR that also targets 4.0 to put the resource group in the zookeeper path. It may make sense to create the other PR for the 4.0 version now (the one that reorganizes the zookeeper path), and merge that before this PR.

Reply:
> I think to do this efficiently, the resource group would need to be in the zookeeper path for all of the server processes (well, the ones that use them).

Created #4861 for this. If that gets merged, then I think the ServiceLockPaths class in that PR can replace the ServerIds class in this PR.

Reply:
Resolving this comment, #4861 was merged.